Source/CarthageKit/DiagnosticResolver.swift

```diff
@@ -0,0 +1,170 @@
+				case .gitReference(let hash):
+					pinnedVersionsProducer = resolvedGitReference(dependency, hash)
+					gitReference = hash
+				default:
+					pinnedVersionsProducer = versionsForDependency(dependency)
+				}
+
+				let concreteVersionsProducer = pinnedVersionsProducer.filterMap { pinnedVersion -> PinnedVersion? in
+					pinnedVersions.append(pinnedVersion)
+					return nil
+				}
+
+				_ = try concreteVersionsProducer.collect().first()!.dematerialize()
+
+				versionSets[dependency] = pinnedVersions
+
+				let storedDependency = self.dependencyMappings?[dependency] ?? dependency
+				try localRepository?.storePinnedVersions(pinnedVersions, for: storedDependency, gitReference: gitReference)
+
+				versionSet = pinnedVersions
+			}
+
+			let filteredVersionSet = versionSet.filter { pinnedVersion -> Bool in
+				versionSpecifier.isSatisfied(by: pinnedVersion)
+			}
+
+			delegate?.diagnosticResolver(self, didFindVersions: filteredVersionSet, compatibleWith: versionSpecifier, for: dependency)
+
+			return filteredVersionSet
+		} catch let error {
+			delegate?.diagnosticResolver(self, didEncounterError: error, whileRetrievingVersionsFor: dependency, compatibleWith: versionSpecifier)
+
+			if ignoreErrors {
+				return [PinnedVersion]()
+			} else {
+				throw error
+			}
+		}
+	}
+
+	private func findDependencies(for dependency: Dependency, version: PinnedVersion) throws -> [(Dependency, VersionSpecifier)] {
+		do {
+			let transitiveDependencies: [(Dependency, VersionSpecifier)] = try dependenciesForDependency(dependency, version).collect().first()!.dematerialize()
+
+			let storedDependency = self.dependencyMappings?[dependency] ?? dependency
+			let storedTransitiveDependencies = transitiveDependencies.map { (transitiveDependency, versionSpecifier) -> (Dependency, VersionSpecifier) in
+				let storedTransitiveDependency = self.dependencyMappings?[transitiveDependency] ?? transitiveDependency
+				return (storedTransitiveDependency, versionSpecifier)
+			}
+			try localRepository?.storeTransitiveDependencies(storedTransitiveDependencies, for: storedDependency, version: version)
+
+			delegate?.diagnosticResolver(self, didFindTransitiveDependencies: transitiveDependencies, for: dependency, withVersion: version)
+
+			return transitiveDependencies
+		} catch let error {
+			delegate?.diagnosticResolver(self, didEncounterError: error, whileRetrievingTransitiveDependenciesFor: dependency, withVersion: version)
+
+			if ignoreErrors {
+				return [(Dependency, VersionSpecifier)]()
+			} else {
+				throw error
+			}
+		}
+	}
+
+	private struct PinnedDependency: Hashable {
+		public let dependency: Dependency
+		public let pinnedVersion: PinnedVersion
+		private let hash: Int
+
+		init(dependency: Dependency, pinnedVersion: PinnedVersion) {
+			self.dependency = dependency
+			self.pinnedVersion = pinnedVersion
+			self.hash = 37 &* dependency.hashValue &+ pinnedVersion.hashValue
+		}
+
+		public var hashValue: Int {
```

Comment:
How about making use of synthesizing hashable conformance?

Reply:
That only works for Xcode 9.3 or later (Swift 4.1) but we still support Xcode 9.2 (Swift 4.0).

Once we migrate to ReactiveSwift 4.0 which requires Swift 4.1, we can rely on that feature.

Reply:
I have a WIP branch locally to update the dependencies so stay tuned :wink:

Reply:
Aha, got it.

Reply:
A problem with relying on auto-synthesis of hashable is performace. The hashValue should only be calculated once in the initializer and not every time hashValue is called. This actually is a big improvement in performance I did in my other pull request with the fast resolver.